src/Diagnostics.RuntimeHost/Services/SourceWatcher/Watchers/GitHubWatcher.cs

```diff
@@ -1,105 +1,146 @@
+
+        // Load from configuration.
         private string _destinationCsxPath;
         private int _pollingIntervalInSeconds;
 
         protected override Task FirstTimeCompletionTask => _firstTimeCompletionTask;
 
         protected override string SourceName => "GitHub";
 
-        public GitHubWatcher(IHostingEnvironment env, IConfiguration configuration, IInvokerCacheService invokerCache, IGithubClient githubClient)
-            : base(env, configuration, invokerCache, "GithubWatcher")
+        private IDictionary<string, IGithubWorker> GithubWorkers { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GitHubWatcher" /> class.
+        /// </summary>
+        /// <param name="env">Hosting environment.</param>
+        /// <param name="configuration">The configuration.</param>
+        /// <param name="invokerCache">Invoker cache.</param>
+        /// <param name="gistCache">Gist cache.</param>
+        /// <param name="githubClient">Github client.</param>
+        public GitHubWatcher(IHostingEnvironment env, IConfiguration configuration, IInvokerCacheService invokerCache, IGistCacheService gistCache, IGithubClient githubClient)
+            : base(env, configuration, invokerCache, gistCache, "GithubWatcher")
         {
             _githubClient = githubClient;
 
-            _lastModifiedMarkerName = "_lastModified.marker";
-            _deleteMarkerName = "_delete.marker";
-            _cacheIdFileName = "cacheId.txt";
             LoadConfigurations();
+
+            #region Initialize Github Worker
+
+            // TODO: Register the github worker with destination path.
+            var gistWorker = new GithubGistWorker(gistCache);
+            var detectorWorker = new GithubDetectorWorker(invokerCache);
+
+            GithubWorkers = new Dictionary<string, IGithubWorker>
+            {
+                { gistWorker.Name, gistWorker },
+                { detectorWorker.Name, detectorWorker }
+            };
+
+            #endregion
+
             Start();
         }
 
+        /// <summary>
+        /// Start github watcher.
+        /// </summary>
         public override void Start()
         {
             _firstTimeCompletionTask = StartWatcherInternal();
             StartPollingForChanges();
         }
 
-        public override async Task<Tuple<bool, Exception>> CreateOrUpdateDetector(DetectorPackage pkg)
+        /// <summary>
+        /// Create or update a package.
+        /// </summary>
+        /// <param name="pkg">Detector package.</param>
+        /// <returns>Task for creating or updateing detector.</returns>
+        public override async Task<Tuple<bool, Exception>> CreateOrUpdatePackage(Package pkg)
         {
             if (pkg == null)
             {
-                return new Tuple<bool, Exception>(false, new ArgumentNullException("pkg"));
+                return new Tuple<bool, Exception>(false, new ArgumentNullException(nameof(pkg)));
```

Comment:
Somehow I feel this is over engineering, you can just throw the ArgumentNullException and not have to wrap it inside of some Tuple response.

Reply:
I try to reduce the chagne of code. I will remove the redundant code.